source/agora/consensus/data/Block.d

```diff
@@ -775,8 +775,25 @@ unittest
 version (unittest)
 {
     import agora.utils.Log;
+    import agora.crypto.ECC;
+
     mixin AddLogger!();
 
+    Height[PublicKey] first_enrolled;
+
+    // Note that the height here is for the next block to be externalized
+    Hash getPreimageForValidator (in KeyPair key_pair, in Height height) @safe nothrow
+    {
+        if ((key_pair.address in first_enrolled) is null)
+            first_enrolled[key_pair.address] = height;
+        auto times = 1999 - height - first_enrolled[key_pair.address];
+        log.trace("getPreimageForValidator: First enrollment for {} was at height {}. hashing {} times", key_pair.address, height, times);
+        auto seed = hashMulti(key_pair.secret, "consensus.preimages", 0);
+        foreach (i;  0 .. times)
+            seed = hashFull(seed);
+        return seed;
+    }
```

Comment:
Why do we need this ?

Reply:
It is test code to provide the pre images for the unit tests in `node.Ledger` and `data.Block`.

Reply:
Removed by merged PR: https://github.com/bosagora/agora/pull/2052